### PR TITLE
Fraudproofs for PR

### PR DESCRIPTION
**Pull request by ksloven**
A report on fraud proofs and different implementations thereof on different blockchains

### CODE REVIEW

File: cryptography/fraud-proofs-1/MainReport.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+Fraud proofs are integral to the security of SPV clients, however, the other components in SPV clients are not without issues. 
+
+
+## Security and privacy issues with SPV clients
+* **weak bloom filters and merkle tree designs**
+
+In August 2017, a weakness in the Bitcoin Merkle tree design was found to reduce the security of SPV clients which could allow an attacker to simulate a payment of arbitrary amount to a victim using a SPV wallet, and trick the victim into accepting it as valid [10]. The bitcoin Merkle tree makes no distinction between inner and leaf nodes and could thus be manipulated by an attack that could re-interpret transactions as nodes and nodes as transactions [11]. This weakness is due to inner nodes having no format and only requiring the length to be 64 bytes.
+
+This brute force attack particularly affects systems that automatically accept SPV proofs and could be carried out with an investment of approximately $3 million.One proposed solution is to ensure that no internal, 64-bit node is ever accepted as a valid transaction by SPV wallets/clients [11].
```
[the remark this answers is not in the document]
Comment (ksloven) [reply cryptography/fraud-proofs-1/MainReport.md]: amended

File: cryptography/fraud-proofs-1/MainReport.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+Furthermore, SPV clients pose the risk of a denial of service attack against full nodes due to processing load (80Gig disk reads) when SPV clients sync and full nodes themselves can cause a denial of service against SPV clients by returning NULL filter responses to requests [14]. Peter Todd's Bloom-io-attack aptly demonstrates the risk of SPV denial of service [15].
+
+To address this, a new concept called committed bloom filters was introduced to improve the performance and security of SPV clients. In this concept, which can be used in lieu of BIP37 [16], a Bloom filer digest (BFD) of every blocks inputs, outputs and transactions is created with a filter that consists of a small size of the overall block size [14].
+
+A second Bloom filter is created with all transactions and a binary comparison is made to determine matching transactions. This BFD allows the caching of filters by SPV clients without the need to re-compute [16] and also introduces semi-trusted oracles to improve the security and privacy of SPV clients by allowing SPV clients to download block data via any out of band method. [14]
```
[the remark this answers is not in the document]
Comment (ksloven) [reply cryptography/fraud-proofs-1/MainReport.md]: amended

File: cryptography/fraud-proofs-1/MainReport.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+## What are fraud proofs?
+
+Fraud proofs are a way to improve the security of SPV clients [5] by providing a mechanism for full nodes to prove that a chain is invalid irrespective of the amount of proof of work it has [5]. Fraud proofs could also help with the Bitcoin scaling debate as SPV clients are easier to run and could thus help with Bitcoin scalability issues [6][18].
+
+A full Bitcoin node contains the following details:
+  * every transaction that is currently being broadcast around the network
+  * every transaction that has ever been sent
+  * all the unspent transaction outputs (UTXOs) [4]
+  
+These SPV client make use of Bloom filters to receive transactions that are relevant to the user [7]. Bloom filters are probalistic data structures used to check the existence of an element in a set quicker by respond with a boolean answer [9]
```
[the remark this answers is not in the document]
Comment (ksloven) [reply cryptography/fraud-proofs-1/MainReport.md]: amended

File: cryptography/fraud-proofs-1/MainReport.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
+# Fraud Proofs - easier said than done?
+
+## Background
+The Bitcoin blockchain is, as of June 2018, approximately 173 Gigabytes in size [1]. This makes it nearly impossible for everyone to run a full Bitcoin node. Lightweight/Simplified Payment Verification (SPV) clients will have to be used by users since not everyone can run full nodes due to the computational power and cost needed to run a full Bitcoin node. 
+
+![spv21.png](sources/spv21.png)
+Courtesy:MIT Bitcoin Expo 2016 Day 1
```
[the remark this answers is not in the document]
Comment (ksloven) [reply cryptography/fraud-proofs-1/MainReport.md]: done

File: cryptography/fraud-proofs-1/MainReport.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+   - [What are fraud proofs](#what-are-fraud-proofs)
+   - [Security and privacy issues with SPV clients](#security-and-privacy-issues-with-spv-clients)
+   - [Examples of SPV implementations](#examples-of-spv-implementations)
+   - [Fraud proof implementations in other blockchains](#fraud-proof-implementations-in-other-blockchains)
+   - [Suggested fraud proof improvements](#suggested-fraud-proof-improvements)
+   - [Conclusions, Observations, Recommendations](#conclusions-observations-recommendations)
+   - [References](#references)
+
+In the original Bitcoin whitepaper, Satoshi recognised this and introduced the concept of a Simplified Payment Verification (SPV) [2], in which he describes a technique that allows verification of payments using a lightweight client that doesn't need to download the entire Bitcoin blockchain, but rather by only downloading block headers with the longest proof-of-work chain , which are achieved by obtaining the Merkle branch linking a transaction to a block[3]. The existence of Merkle root in the chain, along with blocks added after the block containing the Merkle root, provides confirmation of the legitimacy of that chain.
```
[the remark this answers is not in the document]
Comment (ksloven) [reply cryptography/fraud-proofs-1/MainReport.md]: amended

File: cryptography/fraud-proofs-1/MainReport.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
+# Fraud Proofs - easier said than done?
+
+## Background
+The Bitcoin blockchain is, as of June 2018, approximately 173 Gigabytes in size [1]. This makes it nearly impossible for everyone to run a full Bitcoin node. Lightweight/Simplified Payment Verification (SPV) clients will have to be used by users since not everyone can run full nodes due to the computational power and cost needed to run a full Bitcoin node. 
+
```
[the remark this answers is not in the document]
Comment (ksloven) [reply cryptography/fraud-proofs-1/MainReport.md]: amended

**Comment by hansieodendaal** [comment]
Hi @ksloven , references comment of issue #65 still outstanding.

- [ ] References: Use internal links for references and enable opening of URLs at the same time. This dead reference (**\[1\]**) in the text will become this 'live' reference (**[[1]]**) that is clickable and with data tool tip shown. See example below.

[1]: https://www.statista.com/statistics/647523/worldwide-bitcoin-blockchain-size 
"Size of the Bitcoin blockchain from 2010 to 2018, 
by quarter (in megabytes)"

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Referenced in the text like this:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`[[1]]`

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Include reference like this:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`[[1]]<space>Title,<space>Author (if paper),<space>URL,<space>Date accessed.`

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`[1]:<space>URL<space>`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`"Title line 1`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`Title line n"`

Regards

File: cryptography/fraud-proofs-1/MainReport.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
+# Fraud Proofs - easier said than done?
+
+## Background
+The Bitcoin blockchain is, as of June 2018, approximately 173 Gigabytes in size [1]. This makes it nearly impossible for everyone to run a full Bitcoin node. Lightweight/Simplified Payment Verification (SPV) clients will have to be used by users since not everyone can run full nodes due to the computational power and cost needed to run a full Bitcoin node.
```
[the remark this answers is not in the document]
Comment (ksloven) [reply cryptography/fraud-proofs-1/MainReport.md]: done

File: cryptography/fraud-proofs-1/MainReport.md [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+
+In this system, the full nodes would need to provide an alert (known as a fraud proof) to SPV clients when an invalid block is detected. The SPV clients would then be prompted to download the full block and alerted transactions to
+confirm the inconsistency [2].
+
+An invalid block need not be of malicious intent, but could be as a result of any of the following [6]:
+* **Bad Txn** (invalid txn, doublespent txn, or repeat txn).
+* **Missing block data** (unknown and undiscoverable Merkle trees – this could be intentional or accidental).
+* **Bad Block** (Other) (misplaced coinbase, wrong version, witness data missing, (drivechain) most updates to Escrow_DB/Withdrawal_DB)
+* **Bad Accumulation** (the infamous blocksize/SigOps limits, the coinbase txn fees (which must balance total fees paid by the block’s txns), (drivechain) sidechain outputs – the “CTIP” field of “Escrow DB”)
```
[the remark this answers is not in the document]
Comment (ksloven) [reply cryptography/fraud-proofs-1/MainReport.md]: done

**Comment by hansieodendaal** [comment]
@ksloven please see further comments:

- [ ] Update **Contents** with all headings

- [ ] Add new heading just after **Contents** before '_In the original Bitcoin whitepaper_'

- [ ] Please add references for 'https://github.com/petertodd/bitcoin/commit/53dda4643a0dba85699cc4b9dc06a0b9afb1d315' and 'https://github.com/petertodd/bitcoin/tree/2016-02-lie-to-spv' and ''

- [ ] Please find more original reference for Merkle root, remove faulty reference [Merkle root](#blob/master/merkle-trees-and-spv-1/PITCHME.md#merkle-trees-and-more-importantly-branches) 

Thanks!

**Comment by hansieodendaal** [comment]
@CjS77 can you approve Kenny's changes if happy? We need to squash merge this branch to master so his images can be referenced from master for his GitPich; he will still develop his presentation on this branch

**Comment by CjS77** [comment]
Nicely done @ksloven

**Pull request merged**